presto-parquet/src/main/java/io/prestosql/parquet/reader/ShortDecimalColumnReader.java

```diff
@@ -47,54 +52,104 @@
+            DecimalType prestoDecimalType;
+
+            if (prestoType instanceof DecimalType) {
+                prestoDecimalType = (DecimalType) prestoType;
             }
-            else if (prestoType.equals(TINYINT) || prestoType.equals(SMALLINT) || prestoType.equals(INTEGER) || prestoType.equals(BIGINT)) {
+            else {
+                // for numeric types convert decimal to unscaled long
+                prestoDecimalType = DecimalType.createDecimalType(MAX_SHORT_PRECISION, 0);
+
                 if (parquetDecimalType.getScale() != 0) {
-                    throw new ParquetDecodingException(format(
-                            "Parquet decimal column type with non-zero scale (%s) cannot be converted to Presto %s column type",
-                            parquetDecimalType.getScale(),
-                            prestoType));
+                    throw new PrestoException(NOT_SUPPORTED, format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
                 }
-                prestoType.writeLong(blockBuilder, decimalValue);
             }
-            else {
-                throw new ParquetDecodingException(format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
+
+            long rescale = longTenToNth(Math.abs(prestoDecimalType.getScale() - parquetDecimalType.getScale()));
+
+            if (isShortDecimal(prestoDecimalType)) {
+                long scaledValue = shortToShortCast(
+                        value,
+                        parquetDecimalType.getPrecision(),
+                        parquetDecimalType.getScale(),
+                        prestoDecimalType.getPrecision(),
+                        prestoDecimalType.getScale(),
+                        rescale,
+                        rescale / 2);
+
+                prestoType.writeLong(blockBuilder, scaledValue);
+            }
+            else if (isLongDecimal(prestoDecimalType)) {
+                prestoType.writeSlice(blockBuilder, shortToLongCast(
+                        value,
+                        parquetDecimalType.getPrecision(),
+                        parquetDecimalType.getScale(),
+                        prestoDecimalType.getPrecision(),
+                        prestoDecimalType.getScale()));
+            }
+            else if (isNumberType(prestoType)) {
+                if (isInValidNumberRange(prestoType, value)) {
+                    prestoType.writeLong(blockBuilder, value);
+                }
+                else {
+                    throw new PrestoException(NOT_SUPPORTED, format("Could not coerce from %s to %s", parquetDecimalType, prestoType));
+                }
             }
         }
         else if (isValueNull()) {
             blockBuilder.appendNull();
         }
     }
 
-    private void validateDecimal(DecimalType prestoType)
+    protected boolean isNumberType(Type type)
```

Comment:
- `private` (here & below)
- REAL, DOUBLE and DECIMAL are also number types

Reply:
protected